src/lib.rs

```diff
@@ -19,55 +50,282 @@ pub mod common {
+                    system_id: sysid,
+                    component_id: compid,
+                },
+                msg,
+            ));
         }
     }
 }
 
-/// Write a MAVLink message to a Write stream.
-pub fn write<W: Write>(w: &mut W, header: Header, data: &MavMessage) -> io::Result<()> {
+#[cfg(feature="mavlink2")]
+const MAVLINK_IFLAG_SIGNED: u8 = 0x01;
+
+///
+/// Read a MAVLink v2  message from a Read stream.
+#[cfg(all(feature = "std", feature="mavlink2"))]
+pub fn read_msg<R: Read>(r: &mut R) -> Result<(MavHeader, MavMessage)> {
+    loop {
+        // search for the magic framing value indicating start of mavlink message
+        if r.read_u8()? != MAV_STX_V2 {
+            continue;
+        }
+
+//        println!("Got STX2");
+        let payload_len = r.read_u8()? as usize;
+//        println!("Got payload_len: {}", payload_len);
+        let incompat_flags = r.read_u8()?;
+//        println!("Got incompat flags: {}", incompat_flags);
+        let compat_flags = r.read_u8()?;
+//        println!("Got compat flags: {}", compat_flags);
+
+        let seq = r.read_u8()?;
+//        println!("Got seq: {}", seq);
+
+        let sysid = r.read_u8()?;
+//        println!("Got sysid: {}", sysid);
+
+        let compid = r.read_u8()?;
+//        println!("Got compid: {}", compid);
+
+        let mut msgid_buf = [0;4];
+        msgid_buf[0] = r.read_u8()?;
+        msgid_buf[1] = r.read_u8()?;
+        msgid_buf[2] = r.read_u8()?;
+
+        let header_buf = &[payload_len as u8,
+            incompat_flags, compat_flags,
+            seq, sysid, compid,
+            msgid_buf[0],msgid_buf[1],msgid_buf[2]];
+
+        let msgid: u32 = unsafe { transmute(msgid_buf) };
+//        println!("Got msgid: {}", msgid);
+
+        //provide a buffer that is the maximum payload size
+        let mut payload_buf = [0; 255];
+        let payload = &mut payload_buf[..payload_len];
+
+        r.read_exact(payload)?;
+
+        let crc = r.read_u16::<LittleEndian>()?;
+
+        if (incompat_flags & 0x01) == MAVLINK_IFLAG_SIGNED {
+            let mut sign = [0;13];
+            r.read_exact(&mut sign)?;
+        }
+
+        let mut crc_calc = crc16::State::<crc16::MCRF4XX>::new();
+        crc_calc.update(header_buf);
+        crc_calc.update(payload);
+        let extra_crc = MavMessage::extra_crc(msgid);
+
+        crc_calc.update(&[extra_crc]);
+        let recvd_crc = crc_calc.get();
+        if recvd_crc != crc {
+//            println!("msg id {} payload_len {} , crc got {} expected {}", msgid, payload_len, crc, recvd_crc );
+            continue;
+        }
+
+        if let Some(msg) = MavMessage::parse(msgid, payload) {
+            return Ok((
+                MavHeader {
+                    sequence: seq,
+                    system_id: sysid,
+                    component_id: compid,
+                },
+                msg,
+            ));
+        }
+        else {
+            println!("invalid MavMessage::parse");
```

Comment:
Maybe return ErrorKind::InvalidData. Library code in general shouldn't just print its own error messages. Let the program using this library control its UI.